plugins/woocommerce/src/Internal/Orders/SourceAttributionBlocksController.php

```diff
@@ -0,0 +1,237 @@
+	 * @param SourceAttributionController $source_attribution_controller Instance of the source attribution controller.
+	 */
+	final public function init(
+
+		/*
+		AssetApi $asset_api,
+		*/
+		ExtendSchema $extend_schema,
+		FeaturesController $features_controller,
+		SourceAttributionController $source_attribution_controller
+	) {
+		// $this->asset_api                     = $asset_api;
+		$this->extend_schema                 = $extend_schema;
+		$this->features_controller           = $features_controller;
+		$this->source_attribution_controller = $source_attribution_controller;
+	}
+
+	/**
+	 * Hook into WP.
+	 *
+	 * @since x.x.x
+	 * @return void
+	 */
+	public function register() {
+		// Bail if the feature is not enabled.
+		if ( ! $this->features_controller->feature_is_enabled( 'order_source_attribution' ) ) {
+			return;
+		}
+
+		$this->extend_api();
+
+		// Bail early on admin requests to avoid asset registration.
+		if ( is_admin() ) {
+			return;
+		}
+
+		add_action(
+			'init',
+			function() {
+				$this->register_assets();
+			}
+		);
+
+		add_action(
+			'wp_enqueue_scripts',
+			function() {
+				$this->enqueue_scripts();
+			}
+		);
+	}
+
+	/**
+	 * Register scripts.
+	 */
+	private function register_assets() {
+		// Once Blocks are moved to the monorepo, hopefully, we can use the following code.
+
+		/*
+		$this->asset_api->register_script(
+			'wc-blocks-order-source-attribution',
+			'build/wc-blocks-order-source-attribution.js',
+			[ 'wc-order-source-attribution' ]
+		);
+		*/
+		// Instead of the elaborate version below.
+		wp_register_script(
+			'wc-order-source-attribution-blocks',
+			plugins_url(
+				"assets/js/frontend/order-source-attribution-blocks{$this->get_script_suffix()}.js",
+				WC_PLUGIN_FILE
+			),
+			array( 'wc-order-source-attribution', 'wp-data', 'wc-blocks-checkout' ),
+			Constants::get_constant( 'WC_VERSION' ),
+			true
+		);
+	}
+
+	/**
+	 * Enqueue the Order Source Attribution script.
+	 *
+	 * @since x.x.x
+	 * @return void
+	 */
+	private function enqueue_scripts() {
+		wp_enqueue_script( 'wc-order-source-attribution-blocks' );
+	}
+
+	/**
+	 * Extend the Store API.
+	 *
+	 * @return void
+	 */
+	private function extend_api() {
+		$this->extend_schema->register_endpoint_data(
+			array(
+				'endpoint'        => CheckoutSchema::IDENTIFIER,
+				'namespace'       => 'woocommerce/order-source-attribution',
+				'schema_callback' => $this->get_schema_callback(),
+			)
+		);
+		// The following does not work.
+		// See https://github.com/woocommerce/woocommerce-blocks/blob/d646b42e98dc69fe06356f54b53fd6cff132fe98/docs/third-party-developers/extensibility/rest-api/extend-rest-api-update-cart.md#basic-usage .
+
+		/*
+		$this->extend_schema->register_update_callback(
+			[
+				'namespace' => 'woocommerce/order-source-attribution',
+				'callback'  => function( $data ) {
+					// Save the data to the order here.
+				},
+			]
+		);
+		*/
```

Comment:
I'd like to understand what's not working about this.

This API is designed to update the WC_Cart - not to modify the order after "Place order" has been pressed. The hook used below is the correct way for this.

Was there some incorrect expectations based on the docs, or code Rubik shared with you that made you try this?

My main concern is giving devs the best DX possible, so any feedback on that point would be great!

Reply:
Thank you for the clarifications! :) 

What made me try this, was that this function was present in https://github.com/woocommerce/woocommerce-blocks/pull/11506 I took over

From the name itself, I read `register_endpoint_data` is to register an endpoint in the API to receive the data, with the schema etc., and `register_update_callback` is to register the callback to update when the data is recieved. 
The name itself does not give a clue on restriction to the cart only. 

Then I searched the code and docs for anything related to `register_endpoint_data` and `register_update_callback`, and found this:
https://github.com/woocommerce/woocommerce-blocks/blob/d646b42e98dc69fe06356f54b53fd6cff132fe98/docs/third-party-developers/extensibility/rest-api/extend-rest-api-update-cart.md#basic-usage which states:
> `ExtendSchema` offers the ability for extensions to register **callback functions to be executed** when signalled to do so **by** the client-side Cart **or Checkout**.

So I thought: " The function name does not seem to be cart-specific; the cart is what will eventually be checked out. Therefore updating the cart should update what will be checked out," - so it made sense to me to use it.
I was a bit unsure, as it is not taking the `'endpoint'` parameter, but I thought maybe it is not needed, and the callback is chosen based on the `namespace` matching.

To double-check, `$this->extend_schema->register_endpoint_data(` is the right way to extend what would be expected on the checkout request?

Reply:
I removed the comment in https://github.com/woocommerce/woocommerce/pull/41592/commits/77889265bccbca59cb752e8f01d86750761df2fa